### PR TITLE
Performance projects for services

### PR DESCRIPTION
**Pull request by vhvb1989**
#1637 needs to be merged before this one

fixes: #1654 
fixes: #1636

### CODE REVIEW

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]
We don't want perf test for datalake or share. That would be extra maintain efforts. We only perform perf test on blob

**Comment by vhvb1989** [comment]
@Jinming-Hu  can you let the project anyway?  As a reference?  It wouldn't be added to any auto test integration or anything, it would be just over there.

If not, that's fine, I can remove it from there

**Comment by Jinming-Hu** [comment]
> @Jinming-Hu can you let the project anyway? As a reference? It wouldn't be added to any auto test integration or anything, it would be just over there.
> 
> If not, that's fine, I can remove it from there

I prefer we just remove it. Nobody is gonna use it anyway

**Review by Jinming-Hu — APPROVED** [review]

**Review by RickWinter — COMMENTED** [review]
Should performance be under test folder?  I see in other languages its a sibling

File: sdk/core/azure-core/test/performance/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+
+# Configure CMake project.
+cmake_minimum_required (VERSION 3.13)
+project(azure-core-performance-test LANGUAGES CXX)
```
Comment (RickWinter) [inline sdk/core/azure-core/test/performance/CMakeLists.txt:6]:
```suggestion
project(azure-core-performance LANGUAGES CXX)
```

File: sdk/core/azure-core/test/performance/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+set(
+  AZURE_CORE_PERF_TEST_SOURCE
+    src/main.cpp
+)
+
+# Name the binary to be created.
+add_executable (
+  azure-core-performance-test
```
Comment (RickWinter) [inline sdk/core/azure-core/test/performance/CMakeLists.txt:22]:
```suggestion
  azure-core-performance
```

File: sdk/core/azure-core/test/performance/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+
+set(
+  AZURE_CORE_PERF_TEST_HEADER
+  inc/azure/core/test/performance/nullable.hpp
+)
+
+set(
+  AZURE_CORE_PERF_TEST_SOURCE
```
Comment (RickWinter) [inline sdk/core/azure-core/test/performance/CMakeLists.txt:16]: formating

File: sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ */
+
+#pragma once
+
+#include <azure/performance_framework.hpp>
+
+#include <memory>
+
+namespace Azure { namespace Core { namespace Test { namespace Performance {
```
Comment (RickWinter) [inline sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp:16]: NIT, namespace Azure::Performance::Core

File: sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  /**
+   * @brief Measure the Nullable object performance.
+   *
+   */
+  class NullableTest : public Azure::PerformanceStress::PerformanceTest {
```
Comment (RickWinter) [inline sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp:22]: I favor "Perf" over PerformanceStress/Performance.  Perf is very common name for Perf testing.  Its short and concise.  If you want to go with Performance that would be my second choice

File: sdk/core/performance-stress/src/program.cpp [hunk 6]
```diff
@@ -14,26 +14,19 @@
...
 
-inline std::function<std::unique_ptr<Azure::PerformanceStress::PerformanceTest>(
-    Azure::PerformanceStress::TestOptions)>
-GetTest(
-    std::map<
-        std::string,
-        std::function<std::unique_ptr<Azure::PerformanceStress::PerformanceTest>(
-            Azure::PerformanceStress::TestOptions)>> const& tests,
+inline Azure::PerformanceStress::TestMetadata const* GetTestMetadata(
+    std::vector<Azure::PerformanceStress::TestMetadata> const& tests,
```
Comment (RickWinter) [inline sdk/core/performance-stress/src/program.cpp:29]: Store the tests as a Map as you had before.  Use the test name as the key, its ok to have the key in the map and in the TestMetadata.
The map is much more efficient for finding the tests we want to run.

File: sdk/core/performance-stress/src/program.cpp [hunk 7]
```diff
@@ -89,7 +82,8 @@ inline void PrintOptions(
     {
       try
       {
-        optionsJs[option.Name] = parsedArgs[option.Name].as<std::string>();
+        optionsJs[option.Name]
```
Comment (RickWinter) [inline sdk/core/performance-stress/src/program.cpp:85]: Nit, optionsJs isn't clear on what it is.  I have to go find it to realize Js was short for json.

**Comment by RickWinter** [comment]
Please fix the packagename, namespace, and file structure before submitting

File: sdk/core/performance-stress/src/program.cpp (hunk 6 above)
Comment (vhvb1989) [reply sdk/core/performance-stress/src/program.cpp:29]: we don't really need that efficiency for finding the test to run. 
I am hoping to keep the perf program not knowing test details, as a test writer, I only want to take care of creating the test without worry too much about how do I add it to the perf program

**Comment by vhvb1989** [comment]
> Should performance be under test folder? I see in other languages its a sibling

Probably we don't have a guideline for test-strategy structure.  I think any kind of testing should live under the same starting directory /test.  We could have
- unit
- performance
- stress
- continuous operation
- functional

I don't see a good reason why to have these folders directly inside the package root folder like
- sdk/core/azure-core
  - test
  - performance
  - stress
  - continuous-operation
  - functional
  - src
  - inc
  - vcpkg

IMHO, if I want to add/review/update any test, I should easily find all of them inside /test as:
- sdk/core/azure-core
  - test
    - unit
    - performance
    - stress
    - continuous-operation
    - functional
  - src
  - inc
  - vcpkg

Then, if I am not interested in any kind of tests, I don't see all of them next to the inc and scr folders (main folders).

I don't have strong opinion about the namespace structure. But updating the namespace structure of the tests is a more all-repo general change. Changing only here would let the repo with 2 approaches. So, if we want to change it, it's better to dedicate one PR only for it.

File: sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp (hunk 4 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp:16]: I will create a new issue to define the test file structure and namespace that we want to follow.
As right now we are not doing Azure::Test::  or Azure::Performance and making this change only for this PR wouldn't be ideal.

Comment (vhvb1989) [reply sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp:16]: https://github.com/Azure/azure-sdk-for-cpp/issues/1694

File: sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp (hunk 5 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/performance/inc/azure/core/test/performance/nullable.hpp:22]: Rename the perf name: https://github.com/Azure/azure-sdk-for-cpp/issues/1696

**Pull request merged**